naga/src/back/spv/block.rs

```diff
@@ -253,6 +258,445 @@ impl Writer {
+                .position(|a| a.binding == crate::Binding::BuiltIn(crate::BuiltIn::VertexCount))
+                .unwrap() as u32,
+        );
+        let prim_count_id_before_max = load_u32_by_member_index(
+            return_info
+                .out_members
+                .iter()
+                .position(|a| a.binding == crate::Binding::BuiltIn(crate::BuiltIn::PrimitiveCount))
+                .unwrap() as u32,
+        );
+
+        // Clamp them to the allowed range
+        let vert_count_id = self.id_gen.next();
+        block.body.push(Instruction::ext_inst(
+            self.gl450_ext_inst_id,
+            spirv::GLOp::UMin as Word,
+            u32_id,
+            vert_count_id,
+            &[vert_count_id_before_max, return_info.max_vertices_constant],
+        ));
+        let prim_count_id = self.id_gen.next();
+        block.body.push(Instruction::ext_inst(
+            self.gl450_ext_inst_id,
+            spirv::GLOp::UMin as Word,
+            u32_id,
+            prim_count_id,
+            &[
+                prim_count_id_before_max,
+                return_info.max_primitives_constant,
+            ],
+        ));
+        // Get pointers to the arrays of data to extract
+        let vert_array_ptr = self.id_gen.next();
+        block.body.push(Instruction::access_chain(
+            self.get_pointer_type_id(
+                return_info.vertex_array_type_id,
+                spirv::StorageClass::Workgroup,
+            ),
+            vert_array_ptr,
+            return_info.out_variable_id,
+            &[self.get_constant_scalar(crate::Literal::U32(
+                return_info
+                    .out_members
+                    .iter()
+                    .position(|a| a.binding == crate::Binding::BuiltIn(crate::BuiltIn::Vertices))
+                    .unwrap() as u32,
+            ))],
+        ));
+        let prim_array_ptr = self.id_gen.next();
+        block.body.push(Instruction::access_chain(
+            self.get_pointer_type_id(
+                return_info.primitive_array_type_id,
+                spirv::StorageClass::Workgroup,
+            ),
+            prim_array_ptr,
+            return_info.out_variable_id,
+            &[self.get_constant_scalar(crate::Literal::U32(
+                return_info
+                    .out_members
+                    .iter()
+                    .position(|a| a.binding == crate::Binding::BuiltIn(crate::BuiltIn::Primitives))
+                    .unwrap() as u32,
+            ))],
+        ));
+
+        // This must be called exactly once before any other mesh outputs are written
+        {
+            let mut ins = Instruction::new(spirv::Op::SetMeshOutputsEXT);
+            ins.add_operand(vert_count_id);
+            ins.add_operand(prim_count_id);
+            block.body.push(ins);
+        }
+
+        // All this for a `for i in 0..num_vertices` lol
+        // This is basically just unzipping an array and copying to many arrays
```

Comment:
Don't really need this commentary in the code :)


```suggestion
        // This is iterating over every returned vertex and splitting
        // it out into the multiple per-output arrays.
```